### ISSUE TITLE
Uncaught Exception: Couldn't find the sort field 'DocumentSort'

### ISSUE THREAD

**Issue by duclet** [silverstripe-labs/silverstripe-dms#70]
I'm not sure what I'm doing wrong but I get this error right after installing this with Silverstripe 3.2 and trying to edit any page. I have the following modules installed:
* blog
* fluent
* gridfieldextensions
* lumberjack
* tagfield

and of course dms.

**Comment by kbrauer** [comment]
+1

**Comment by samthejarvis** [comment]
The solution for me was to downgrade to 3.1.9. Looks like gridfieldextensions doesn't work with 3.2 yet, something to do with ManyManyLists, perhaps belongs_many_many's aren't instances of ManyManyList in 3.2.

**Comment by benmanu** [comment]
I've pushed up a fork that removes the gridfieldextensions `GridFieldOrderableRows` component if the sortablegridfield module is installed as part of the updateCMSFields call, as they both provide very similar functionality - https://github.com/benmanu/silverstripe-dms/tree/pulls/remove-gridfield-sort-component. Not a permanent fix to the 'DocumentSort' error, just a way to bypass it.

**Comment by DrMartinGonzo** [comment]
I too have this problem.

I bypassed it in another way by doing this check in GridFieldOrderableRows.php getSortTable() method. Pretty horrible I know, but it works

```php
if($list instanceof ManyManyList) {
	$extra = $list->getExtraFields();
	$table = $list->getJoinTable();

	// Hacky fix for DMSDocument sorting
	// ( getExtraFields doesn't return anything )
	if($table == 'DMSDocument_Pages') {
		return $table;
	}

	if($extra && array_key_exists($field, $extra)) {
		return $table;
	}
}
```

**Comment by DrMartinGonzo** [comment]
Forget what I said, this breaks saving DocumentSort ID...
Maybe I'll look further into it when I have time, but the problem seems to be in Silverstripe core isn't that right ?

**Comment by tractorcow** [comment]
It does look like the issue is with GridfieldExtensions. Sorting on Many_Many has been a problem in the past, as well.

**Comment by DrMartinGonzo** [comment]
Thank you for your answer.
It seems getExtraFields() returns an empty array now whereas it used to list all extra fields in Silverstripe 3.1.

Anyway, I've been able to get the many_many_extraFields by doing this in GridFieldOrderableRows:
```php
if($list instanceof ManyManyList) {
    $extra = singleton($list->dataClass)->manyManyExtraFields();
}
```
instead of this :
```php
if($list instanceof ManyManyList) {
    $extra = $list->getExtraFields();
}
```

**Comment by jules0x** [comment]
I confirmed (probably needlessly) this issue exits on a fresh test site built with the latest CWP recipe, with the silverstripe-dms as the only added module.

Our site depends on this module and we're can't upgrade to 3.2 until this has been resolved.

Has any progress been made on a fix for this? (without having to make custom changes)

**Comment by kbrauer** [comment]
Witam serdecznie,
W dniach 25-09/20015 jestem na urlopie bez stałego dostępu do poczty e-mail oraz telefonu.

Na wszystkie maile odpowiem po powrocie. W pilnych sprawach proszę kontaktować się z:

Wojciechem Tyszkowskim
wtyszkowski@wiselimber.pl


Pozdrawiam

Krzysztof Brauer

**Comment by mrmorphic** [comment]
I've written some tests to try to figure this out. I don't believe the core issue is with either dms or gridfieldextensions, although workarounds exist in both. As pointed out above, there are circumstances on a ManyManyList where getExtraFields returns an empty list when there are extra fields defined, and I think that is wrong.

The issue as far as I can tell occurs because DMSDocument has a many_many to SiteTree, and a reverse relationship via the extension DMSSiteTreeExtension. When it gets to DataObject::manyManyExtraFieldsForComponent, there is a test `if($relatedClass === $this->class)` which fails, because relatedClass is SiteTree (defined in the many-many), but the data object instance of course is a Page.

I believe that replacing that test with: `if (is_a($this, $relatedClass))` is better; if the relationship is defined on SiteTree, then it should be found on any sub-class as well. I'm currently running framework test with this change, I'll provide an update if that works.

Note this is reproducible without either dms or gridfieldextensions. I can post the test if anyone is interested.

Other workarounds that appear to work in the dms case are:

 * commenting out the extraFields test in GridFieldOrderableRows::getSortTable; it's only used to double check. If there were a problem and this test is absent, you'd get an error anyway. I've verified this fix works on our site for the DMS case.
 * change DMSDocument Pages relationship to 'Page', and apply DMSSiteTreeExtension to Page instead of SiteTree (in that module's config.) I haven't test that.

M

**Comment by mrmorphic** [comment]
Following up on the above candidate fix, the unit tests all pass, so it appears to not have an adverse affect on anything.

M

**Comment by jules0x** [comment]
@tractorcow appreciate your advice on this one... if this is a framework related problem as identified by @mrmorphic, where do we go from here?

**Comment by tractorcow** [comment]
Make a pull request to framework with the fix and tests? :P

Honestly, if I had more time, I'd get my head into this and help; Unfortunately [we aren't really providing support for this module](http://www.silverstripe.org/software/addons/silverstripe-commercially-supported-module-list/), so it's quite low on the priority list for me.

**Comment by mrmorphic** [comment]
Ok cool. I'll create a PR on framework and update this thread when that's there.

**Comment by mrmorphic** [comment]
Created a PR to framework for this issue https://github.com/silverstripe/silverstripe-framework/pull/4996.

**Comment by mrmorphic** [comment]
Final PR for this issue is https://github.com/silverstripe/silverstripe-framework/pull/5010 which has been merged into framework 3.2.
